### DISCUSSION TITLE
Apple Silicon Support

### DISCUSSION THREAD

**Discussion by shalte1981** [neonbjb/tortoise-tts#115] [General]
Hi! Any plans to add Apple Silicon support? I'm just a basic C# programmer professionally, but I gave it a good try. I had to compile torch audio from source to get tortoise-tts running natively on my M1 Max with GPU support. Unfortunately, it seems that the M1 Max GPU (or specifically the "mps" device equivalent to "cuda" in torch) doesn't support ... something over my head. I got errors about fft ops not being supported.
I don't know how involved this would be, but if it could be made to work it seems like the unified memory architecture in Apple Silicon could be an advantage. Thanks!

<img width="1356" alt="tortoise-tts Native on Apple Silicon" src="https://user-images.githubusercontent.com/98073209/174910289-71a9cd47-a0ed-4bdd-ba4e-d04aaf476a01.png">

**Comment by jnordberg** [comment]
I've also looked into this and came back with the impression the MPS backend is not really ready yet. There's an issue here tracking supported operations https://github.com/pytorch/pytorch/issues/77764

**Reply by neonbjb** [reply]
Whoops wrong heading. :)

**Comment by neonbjb** [comment]
You could just do the ffts on cpu (or use precompiled conditioning latents). There is a section on the latter in the docs. I believe that is the only step that computes an fft.

The rest of the model is traditional matrix math. It should probably work. The only other thing I'd be concerned about is memory consumption. Id love to hear how fast it is if you get it working.

**Reply by shalte1981** [reply]
Cool, thanks for the feedback! This project is amazing and I have a use, so I'll probably go to the cloud in the meantime, but I'll keep an eye on the MPS op coverage tracking issue you posted and keep tinkering with Apple Silicon support 🤓

**Comment by shalte1981** [comment]
I'm going to accept your "MPS backend is not really ready yet" feedback and walk away from Apple Silicon support for now 😂

<img width="1645" alt="Screen Shot 2022-06-22 at 4 58 54 PM" src="https://user-images.githubusercontent.com/98073209/175135314-878700b0-1cb4-4d68-8e32-fbf8307092f5.png">

**Reply by neonbjb** [reply]
LOL.

**Reply by TanishSurana** [reply]
After a year it's still the same :'), why pytorch why!

**Comment by HatimDa** [comment]
> Hi! Any plans to add Apple Silicon support? I'm just a basic C# programmer professionally, but I gave it a good try. I had to compile torch audio from source to get tortoise-tts running natively on my M1 Max with GPU support. Unfortunately, it seems that the M1 Max GPU (or specifically the "mps" device equivalent to "cuda" in torch) doesn't support ... something over my head. I got errors about fft ops not being supported. I don't know how involved this would be, but if it could be made to work it seems like the unified memory architecture in Apple Silicon could be an advantage. Thanks!
> 
> <img alt="tortoise-tts Native on Apple Silicon" width="1356" src="https://user-images.githubusercontent.com/98073209/174910289-71a9cd47-a0ed-4bdd-ba4e-d04aaf476a01.png">

Any news on "mps" support? 
Pytorch has supported apple silicon machines to run on gpu, if someone fixed this let me know.

**Reply by jnordberg** [reply]
Try it, it might work with pytorch nightly now

**Reply by HatimDa** [reply]
> Try it, it might work with pytorch nightly now

I already tried it before posting the question.

It didn't work, hence why I'm asking. Maybe someone found a way to make it work.

**Reply by jnordberg** [reply]
On nightly? What error are you getting?

**Reply by HatimDa** [reply]
Yes on nightly, I installed all the dependencies on my conda environment, everyhting works on cpu side, but it takes so much time, therefore I tried to add an output to see if the model is training on cpu or mps, unfortunately it was using the cpu, even though I tried forcing it to use mps.

Here's the last block of code that I excuted to see if it's mps or cpu:

<img width="706" alt="CleanShot 2023-02-15 at 13 11 19@2x" src="https://user-images.githubusercontent.com/64011336/219024085-bd6305c3-e57d-4473-a996-8254e1327822.png">

When I run verification code from another notebook it works fine!

![CleanShot 2023-02-15 at 15 02 03@2x](https://user-images.githubusercontent.com/64011336/219048896-61038df8-43ff-42f8-8dd3-ec0a68270e87.png)



This makes me guess that we need to change the code inside the tortoise api and swap it from cuda to mps too to make this work.

**Reply by HatimDa** [reply]
Do anyone have a version working on his mac arm machine?

**Comment by takoyaro** [comment]
No luck here either. `mps` support would be great

**Comment by harshitkrverma** [comment]
Is there any update on this? It has been over a year.

**Reply by TanishSurana** [reply]
Nope tried it a week ago still got metal not supported for a attn:? function

**Comment by ErikBjare** [comment]
Merged in https://github.com/neonbjb/tortoise-tts/pull/550!
